tests/test_protocols/test_dialogue/test_base.py

```diff
@@ -1810,22 +1812,171 @@ def test_dump_restore(self):
+        dialogues_storage._skill_component = self.skill_component
+        self.dialogues._dialogues_storage = dialogues_storage
+        dialogues_storage._incomplete_to_complete_dialogue_labels[
+            self.dialogue_label
+        ] = self.dialogue_label
+        self.dialogues.create(
+            self.opponent_address, DefaultMessage.Performative.BYTES, content=b"Hello"
+        )
+        msg, dialogue = self.dialogues.create(
+            self.opponent_address, DefaultMessage.Performative.BYTES, content=b"Hello2"
+        )
+        dialogue.reply(
+            target_message=msg,
+            performative=DefaultMessage.Performative.ERROR,
+            error_code=ErrorCode.UNSUPPORTED_PROTOCOL,
+            error_msg="oops",
+            error_data={},
         )
+        assert dialogues_storage.dialogues_in_terminal_state
+        assert dialogues_storage.dialogues_in_active_state
+        assert dialogues_storage._dialogue_by_address
+        assert dialogues_storage._incomplete_to_complete_dialogue_labels
+        dialogues_by_addr = dialogues_storage.get_dialogues_with_counterparty(
+            dialogue.dialogue_label.dialogue_opponent_addr
+        )
+        dialogues_storage.teardown()
+
+        dialogues_storage_restored = PersistDialoguesStorageWithOffloading(
+            self.dialogues
+        )
+        dialogues_storage_restored._skill_component = self.skill_component
+        dialogues_storage_restored.setup()
+
+        assert len(dialogues_storage._dialogue_by_address) == len(
+            dialogues_storage_restored._dialogue_by_address
+        )
+
         assert (
             dialogues_storage._incomplete_to_complete_dialogue_labels
             == dialogues_storage_restored._incomplete_to_complete_dialogue_labels
         )
+        assert set(
+            [str(i.dialogue_label) for i in dialogues_storage.dialogues_in_active_state]
+        ) == set(
+            [
+                str(i.dialogue_label)
+                for i in dialogues_storage_restored.dialogues_in_active_state
+            ]
+        )
+        assert set(
+            [
+                str(i.dialogue_label)
+                for i in dialogues_storage.dialogues_in_terminal_state
+            ]
+        ) == set(
+            [
+                str(i.dialogue_label)
+                for i in dialogues_storage_restored.dialogues_in_terminal_state
+            ]
+        )
+
+        dialogue_label = dialogues_storage.dialogues_in_terminal_state[0].dialogue_label
+
+        assert len(dialogues_by_addr) == len(
+            dialogues_storage_restored.get_dialogues_with_counterparty(
+                dialogue.dialogue_label.dialogue_opponent_addr
+            )
+        )
+        # check get and cache
+        assert not dialogues_storage_restored._terminal_state_dialogues_labels
+        assert dialogues_storage_restored.get(dialogue_label)
+        assert dialogues_storage_restored._terminal_state_dialogues_labels
```

Comment:
I'd prefer the asserts to be more explicit!

Reply:
slightly updated